src/Microsoft.AspNet.Mvc.TagHelpers/Internal/GlobbingUrlBuilder.cs

```diff
@@ -0,0 +1,122 @@
+
+        /// <summary>
+        /// Builds a list of URLs.
+        /// </summary>
+        /// <param name="staticUrl">The statically declared URL. This will always be added to the result.</param>
+        /// <param name="includePattern">The file globbing include pattern.</param>
+        /// <param name="excludePattern">The file globbing exclude pattern.</param>
+        /// <returns>The list of URLs</returns>
+        public virtual IEnumerable<string> BuildUrlList(string staticUrl, string includePattern, string excludePattern)
+        {
+            var urls = new HashSet<string>(StringComparer.Ordinal);
+
+            // Add the statically declared url if present
+            if (staticUrl != null)
+            {
+                urls.Add(staticUrl);
+            }
+
+            // Add urls that match the globbing patterns specified
+            var matchedUrls = ExpandGlobbedUrl(includePattern, excludePattern);
+            foreach (var url in matchedUrls)
+            {
+                urls.Add(url);
+            }
+
+            return urls;
+        }
+
+        private IEnumerable<string> ExpandGlobbedUrl(string include, string exclude = null)
+        {
+            if (string.IsNullOrEmpty(include))
+            {
+                return Enumerable.Empty<string>();
+            }
+
+            var includePatterns = include.Split(PatternSeparator, StringSplitOptions.RemoveEmptyEntries);
+            var excludePatterns = exclude?.Split(PatternSeparator, StringSplitOptions.RemoveEmptyEntries);
+
+            if (includePatterns.Length == 0)
+            {
+                return Enumerable.Empty<string>();
+            }
+            
+            if (Cache != null)
+            {
+                var cacheKey = $"{nameof(GlobbingUrlBuilder)}-inc:{include}-exc:{exclude}";
+                return Cache.GetOrSet(cacheKey, cacheSetContext =>
+                {
+                    foreach (var pattern in includePatterns)
+                    {
+                        var trigger = FileProvider.Watch(pattern);
+                        cacheSetContext.AddExpirationTrigger(trigger);
+                    }
+
+                    return FindFiles(includePatterns, excludePatterns);
+                });
+            }
+
+            return FindFiles(includePatterns, excludePatterns);
+        }
+
+        private IEnumerable<string> FindFiles(IEnumerable<string> includePatterns, IEnumerable<string> excludePatterns)
+        {
+            var matcher = new Matcher();
+            matcher.AddPatterns(includePatterns, excludePatterns);
+            var matches = matcher.Execute(_baseGlobbingDirectory);
+
+            return matches.Files.Select(ResolveMatchedPath);
+        }
+
+        private string ResolveMatchedPath(string matchedPath)
+        {
+            // Resolve the path to site root
+            var relativePath = new PathString("/" + matchedPath);
+            return RequestPathBase.Add(relativePath).ToString();
```

Comment:
teensy nit: shouldn't need to explicitly create another `PathString` here.  `RequestPathBase + matchedPath` returns the correct `string`.  have you seen cases where that doesn't work?